src/charm.py

```diff
@@ -13,21 +13,35 @@
 from typing import Any, Dict, List, Optional, Union
 
 from charms.grafana_agent.v0.cos_agent import COSAgentRequirer
-from charms.operator_libs_linux.v2 import snap  # type: ignore
+from charms.operator_libs_linux.v2 import snap as snap_lib  # type: ignore
 from charms.tempo_k8s.v1.charm_tracing import trace_charm
 from charms.tempo_k8s.v2.tracing import TracingEndpointRequirer
 from cosl import JujuTopology
 from cosl.rules import AlertRules
 from grafana_agent import METRICS_RULES_SRC_PATH, GrafanaAgentCharm
 from ops.main import main
 from ops.model import BlockedStatus, MaintenanceStatus, Relation
+from snap_management import get_system_arch, install_snap
 
 logger = logging.getLogger(__name__)
 
 _FsType = str
 _MountOption = str
 _MountOptions = List[_MountOption]
 
+_snap_manifest_path = "templates/snap_installation.yaml"
+
+_grafana_agent_snap_spec = {
+    "amd64": {
+        "name": "grafana-agent",
+        "revision": "16",  # 0.35.0
```

Comment:
Bump to 0.39.2 as well to match?

Reply:
My thought is to:
* first land this PR, which gives us a charm revision that is exactly how the current state of the world is.  That way any future updates can restore to it
* following this PR/release, update both amd and arm to 0.40.4 to match [grafana-agent-k8s-operator](https://github.com/canonical/grafana-agent-k8s-operator/blob/66b8a59be154175e5dc34b25fd0b32d84ffeea5f/metadata.yaml#L36)

That way we get all grafana agents on the same workload, but have a downgrade path if something breaks an existing user.  wdyt?

Reply:
Agreed.